<a href="https://colab.research.google.com/github/DShcherbak/University_Labs/blob/main/semester-7/nueral-networks/lectures/lecture_6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Підготовка

In [28]:
from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras import utils
from tensorflow.keras.preprocessing import image
from google.colab import files
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image
%matplotlib inline 

In [29]:
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

In [30]:
classes = ['tshirt', 'pants', 'sweater', 'dress', 'coat', 'shoe', 'shirt', 'sneaker', 'bag', 'boot']

Змінюємо формат даних, нормалізовуємо дані і перетворюємо в формат one hot encoding

In [31]:
x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)

x_train = x_train / 255 
x_test = x_test / 255 

y_train = utils.to_categorical(y_train, 10)
y_test = utils.to_categorical(y_test, 10)

In [52]:
def create_neural(input_layer, hidden_layers, epochs=100, batch_size=200):
  model = Sequential()
  model.add(Dense(input_layer, input_dim=784, activation='relu'))
  for layer in hidden_layers:
    model.add(Dense(layer, activation="relu"))
  model.add(Dense(10, activation="softmax"))
  model.compile(loss="categorical_crossentropy", optimizer="SGD", metrics=["accuracy"])

  

  history = model.fit(
      x_train, y_train, 
      batch_size=batch_size, # 200
      epochs=epochs, # 100
      validation_split=0.2,
      verbose=1
  )

  return model

In [51]:
def check_models(params, epochs=100, batch_size=200):
  results = []
  for param in params:
    in_layer = param[0]
    hidden = param[1]

    model = create_neural(in_layer, hidden, epochs=epochs, batch_size=batch_size)
    score = model.evaluate(x_test, y_test, verbose=1)

    results.append(score[1])
  return results

In [56]:
def print_results(params, results):
  for i in range(len(results)):
    param = params[i]
    result = results[i]
    print("Input layer size:", param[0], end=', ')
    if(len(param[1]) == 0): 
      print("no hidden layers", end='')
    else:
      print("with hidden layers: [", end='')
      for i in range(len(param[1])):
        if(i > 0):
          print(", ", end='')
        print(param[1][i], end='')
      print("]", end=',')  
    print("got result:", f'{result*100:.2f}%')



## Завдання 1. 

Використайте різну кількість нейронів на вхідному шарі: 400, 600, 800, 1200. Аргументуйте відповідь.

In [47]:
params = []
for in_layer in [400,600,800,1200]:
  params.append((in_layer, []))

results = check_models(params)



Epoch 1/100
240/240 [==============================] - 3s 11ms/step - loss: 1.2163 - accuracy: 0.6391 - val_loss: 0.8651 - val_accuracy: 0.7188
Epoch 2/100
240/240 [==============================] - 2s 10ms/step - loss: 0.7851 - accuracy: 0.7481 - val_loss: 0.7150 - val_accuracy: 0.7755
Epoch 3/100
240/240 [==============================] - 2s 10ms/step - loss: 0.6814 - accuracy: 0.7839 - val_loss: 0.6462 - val_accuracy: 0.7908
Epoch 4/100
240/240 [==============================] - 2s 10ms/step - loss: 0.6256 - accuracy: 0.8010 - val_loss: 0.6020 - val_accuracy: 0.8025
Epoch 5/100
240/240 [==============================] - 2s 10ms/step - loss: 0.5890 - accuracy: 0.8103 - val_loss: 0.5745 - val_accuracy: 0.8109
Epoch 6/100
240/240 [==============================] - 2s 10ms/step - loss: 0.5631 - accuracy: 0.8165 - val_loss: 0.5553 - val_accuracy: 0.8157
Epoch 7/100
240/240 [==============================] - 2s 10ms/step - loss: 0.5432 - accuracy: 0.8213 - val_loss: 0.5368 - val_accuracy:

In [50]:
print_results(params, results)

Input layer size: 400, no hidden layersgot result: 86.59%
Input layer size: 600, no hidden layersgot result: 86.81%
Input layer size: 800, no hidden layersgot result: 86.78%
Input layer size: 1200, no hidden layersgot result: 86.64%


Найвища точність вийшла при вхідному шарі розміром 600 нейронів, оскільки подальше збільшення кількості нейронів не збільшувало точності, немає сенсу її збільшувати.

## Завдання 2
Додайте в нейронну мережу прихований шар з різною кількістю
нейронів: 200, 300, 400, 600, 800. Аргументуйте відповідь.

In [53]:
params = []
for hidden in [200,300,400,600,800]:
  params.append((600, [hidden]))

results = check_models(params)

Epoch 1/100
240/240 [==============================] - 5s 18ms/step - loss: 1.2295 - accuracy: 0.6440 - val_loss: 0.8173 - val_accuracy: 0.7417
Epoch 2/100
240/240 [==============================] - 4s 17ms/step - loss: 0.7325 - accuracy: 0.7639 - val_loss: 0.6585 - val_accuracy: 0.7836
Epoch 3/100
240/240 [==============================] - 4s 17ms/step - loss: 0.6247 - accuracy: 0.7992 - val_loss: 0.5879 - val_accuracy: 0.8090
Epoch 4/100
240/240 [==============================] - 4s 17ms/step - loss: 0.5686 - accuracy: 0.8150 - val_loss: 0.5485 - val_accuracy: 0.8158
Epoch 5/100
240/240 [==============================] - 4s 17ms/step - loss: 0.5327 - accuracy: 0.8238 - val_loss: 0.5177 - val_accuracy: 0.8270
Epoch 6/100
240/240 [==============================] - 4s 17ms/step - loss: 0.5090 - accuracy: 0.8306 - val_loss: 0.5015 - val_accuracy: 0.8289
Epoch 7/100
240/240 [==============================] - 4s 17ms/step - loss: 0.4900 - accuracy: 0.8349 - val_loss: 0.4890 - val_accuracy:

In [57]:
print_results(params, results)

Input layer size: 600, with hidden layers: [200],got result: 87.42%
Input layer size: 600, with hidden layers: [300],got result: 87.64%
Input layer size: 600, with hidden layers: [400],got result: 87.70%
Input layer size: 600, with hidden layers: [600],got result: 87.38%
Input layer size: 600, with hidden layers: [800],got result: 87.73%


За результатами 100 епох себе найкраще показали проміжні шари розмірами 300, 400 та 800 нейронів.

## Завдання 3
Додайте кілька прихованих шарів в мережу з різною кількістю
нейронів в кожному шарі. Яким чином це впливає на якість навчання
нейронної мережі.

### Завдання 3.1 
Оскільки перший  додатковий шар ми показав найкращу ефективність при розмірі 800, спробуємо підібрати розмір другого проміжного шару серед розмірів з найкращими показниками (300, 400, 800)

In [60]:
params = []
best_hidden = [800,300,400]
for hid2 in best_hidden:
  params.append((600, [800, hid2]))

results = check_models(params)

Epoch 1/100
240/240 [==============================] - 12s 49ms/step - loss: 1.2782 - accuracy: 0.6405 - val_loss: 0.8073 - val_accuracy: 0.7347
Epoch 2/100
240/240 [==============================] - 11s 48ms/step - loss: 0.7155 - accuracy: 0.7676 - val_loss: 0.6379 - val_accuracy: 0.7933
Epoch 3/100
240/240 [==============================] - 11s 47ms/step - loss: 0.6055 - accuracy: 0.7994 - val_loss: 0.5757 - val_accuracy: 0.8048
Epoch 4/100
240/240 [==============================] - 11s 48ms/step - loss: 0.5503 - accuracy: 0.8135 - val_loss: 0.5291 - val_accuracy: 0.8182
Epoch 5/100
240/240 [==============================] - 12s 49ms/step - loss: 0.5163 - accuracy: 0.8226 - val_loss: 0.5016 - val_accuracy: 0.8268
Epoch 6/100
240/240 [==============================] - 12s 48ms/step - loss: 0.4948 - accuracy: 0.8289 - val_loss: 0.5084 - val_accuracy: 0.8207
Epoch 7/100
240/240 [==============================] - 11s 47ms/step - loss: 0.4791 - accuracy: 0.8336 - val_loss: 0.4773 - val_ac

In [61]:
print_results(params, results)

Input layer size: 600, with hidden layers: [800, 800],got result: 87.37%
Input layer size: 600, with hidden layers: [800, 300],got result: 88.24%
Input layer size: 600, with hidden layers: [800, 400],got result: 88.30%


### Завдання 3.2 
Оптимальна кількість для другого шару - 400, спробуємо додати ще один прихований шар

In [62]:
params = []
for hid3 in best_hidden:
  params.append((600, [800, 400, hid3]))

results = check_models(params)

Epoch 1/100
240/240 [==============================] - 12s 49ms/step - loss: 1.4706 - accuracy: 0.6035 - val_loss: 0.8771 - val_accuracy: 0.7073
Epoch 2/100
240/240 [==============================] - 11s 47ms/step - loss: 0.7627 - accuracy: 0.7419 - val_loss: 0.6772 - val_accuracy: 0.7712
Epoch 3/100
240/240 [==============================] - 12s 48ms/step - loss: 0.6290 - accuracy: 0.7842 - val_loss: 0.5783 - val_accuracy: 0.8033
Epoch 4/100
240/240 [==============================] - 12s 48ms/step - loss: 0.5687 - accuracy: 0.8038 - val_loss: 0.5413 - val_accuracy: 0.8117
Epoch 5/100
240/240 [==============================] - 11s 48ms/step - loss: 0.5287 - accuracy: 0.8152 - val_loss: 0.5157 - val_accuracy: 0.8173
Epoch 6/100
240/240 [==============================] - 12s 48ms/step - loss: 0.5034 - accuracy: 0.8235 - val_loss: 0.4826 - val_accuracy: 0.8322
Epoch 7/100
240/240 [==============================] - 11s 47ms/step - loss: 0.4831 - accuracy: 0.8316 - val_loss: 0.4710 - val_ac

In [63]:
print_results(params, results)

Input layer size: 600, with hidden layers: [800, 400, 800],got result: 88.20%
Input layer size: 600, with hidden layers: [800, 400, 300],got result: 87.73%
Input layer size: 600, with hidden layers: [800, 400, 400],got result: 87.05%


Бачимо, що зростання кількості шарів не несе за собою зростання ефективності.

## Завдання 4
Використовуйте різну кількість епох: 10, 15, 20, 25, 30. Аргументуйте
відповідь.

In [66]:
params = [(600, [800, 400])]
epohs = [10, 15, 20, 25, 30]
results = []

for epoh in epohs:
  result = check_models(params, epochs=epoh)
  results.append(result[0])
  print("===================== (" + str(epoh) + ") =====================")

Epoch 1/10
240/240 [==============================] - 10s 39ms/step - loss: 1.2214 - accuracy: 0.6479 - val_loss: 0.7990 - val_accuracy: 0.7454
Epoch 2/10
240/240 [==============================] - 9s 38ms/step - loss: 0.7127 - accuracy: 0.7673 - val_loss: 0.6388 - val_accuracy: 0.7840
Epoch 3/10
240/240 [==============================] - 9s 38ms/step - loss: 0.6070 - accuracy: 0.7995 - val_loss: 0.5677 - val_accuracy: 0.8087
Epoch 4/10
240/240 [==============================] - 9s 38ms/step - loss: 0.5497 - accuracy: 0.8153 - val_loss: 0.5427 - val_accuracy: 0.8093
Epoch 5/10
240/240 [==============================] - 9s 38ms/step - loss: 0.5170 - accuracy: 0.8242 - val_loss: 0.4999 - val_accuracy: 0.8284
Epoch 6/10
240/240 [==============================] - 9s 38ms/step - loss: 0.4902 - accuracy: 0.8311 - val_loss: 0.5461 - val_accuracy: 0.7927
Epoch 7/10
240/240 [==============================] - 9s 38ms/step - loss: 0.4761 - accuracy: 0.8354 - val_loss: 0.4750 - val_accuracy: 0.832

In [69]:
for i in range(len(epohs)):
  print("Epoh:", epohs[i], "gave the result:", f'{results[i]*100:.2f}%')

Epoh: 10 gave the result: 83.67%
Epoh: 15 gave the result: 84.34%
Epoh: 20 gave the result: 85.30%
Epoh: 25 gave the result: 85.27%
Epoh: 30 gave the result: 86.09%


Зі збільшенням кількості епох зростає точність мережі, однак зростає і час тренування. Після 30 епох точність мережі зростає дуже помірно, тому брати більшу кількість епох сенсу немає.

## Завдання 5
Використовуйте різні розміри міні-вибірки (batch_size): 10, 50, 100, 200, 500. Аргументуйте відповідь.

In [70]:
params = [(600, [800, 400])]
batch_sizes = [500, 200, 100, 50, 10]
results = []

for batch in batch_sizes:
  result = check_models(params, epochs=30, batch_size=batch)
  results.append(result[0])
  print("===================== (" + str(epoh) + ") =====================")

Epoch 1/30
96/96 [==============================] - 8s 82ms/step - loss: 1.7098 - accuracy: 0.5780 - val_loss: 1.2182 - val_accuracy: 0.6739
Epoch 2/30
96/96 [==============================] - 8s 80ms/step - loss: 1.0182 - accuracy: 0.6983 - val_loss: 0.8681 - val_accuracy: 0.7236
Epoch 3/30
96/96 [==============================] - 8s 80ms/step - loss: 0.8071 - accuracy: 0.7472 - val_loss: 0.7424 - val_accuracy: 0.7584
Epoch 4/30
96/96 [==============================] - 8s 79ms/step - loss: 0.7138 - accuracy: 0.7709 - val_loss: 0.6746 - val_accuracy: 0.7742
Epoch 5/30
96/96 [==============================] - 8s 80ms/step - loss: 0.6557 - accuracy: 0.7892 - val_loss: 0.6300 - val_accuracy: 0.7897
Epoch 6/30
96/96 [==============================] - 8s 81ms/step - loss: 0.6158 - accuracy: 0.8009 - val_loss: 0.5946 - val_accuracy: 0.7984
Epoch 7/30
96/96 [==============================] - 8s 80ms/step - loss: 0.5847 - accuracy: 0.8089 - val_loss: 0.5736 - val_accuracy: 0.8064
Epoch 8/30
96

In [71]:
for i in range(len(batch_sizes)):
  print("Batch size:", batch_sizes[i], "gave the result:", f'{results[i]*100:.2f}%')

Batch size: 500 gave the result: 84.26%
Batch size: 200 gave the result: 86.21%
Batch size: 100 gave the result: 86.73%
Batch size: 50 gave the result: 88.13%
Batch size: 10 gave the result: 89.27%


Звісно, зі зменшенням розміру міні-вибірки точність зростає, але швидко зростає і час тренування.

## Підсумок. 
Підберіть різні комбінації гіперпараметрів таким чином, щоб отримати
кращий результат на тестовому наборі даних.
Переконайтеся, що у вашій моделі немає перенавчання.

* Розмір початкового шару: 600
* Розмір першого проміжного шару: 800
* Розмір другого проміжного шару: 400
* Кількість епох: 30
* Розмір міні-вибірки: 10
* Ефективність: 89.27%

